packages/n4s/src/enforce/runner/index.js

```diff
@@ -11,8 +11,17 @@ function runner(rule, value, ...args) {
     return;
   }
 
-  if (rule(value, ...args) !== true) {
-    throw new Error(`[Enforce]: invalid ${typeof value} value`);
+  const ruleResult = rule(value, ...args);
+  // Handles boolean rules
+  if (typeof ruleResult === 'boolean' && ruleResult !== true) {
+    throw new Error(
+      `[Enforce]: invalid ${typeof value} value with rule ${rule.name}`
+    );
+  }
+
+  // Handles object rules { pass, message }
+  if (typeof ruleResult === 'object' && !ruleResult.pass) {
+    throw new Error(`[Enforce]: ${ruleResult.message}`);
   }
 }
```

Comment:
This will fail on null return values because:

```js
typeof null // -> 'object'
```

Also, we did not cover the message function.


I would modify this approach a little to also support the message function. Feel free to make changes to it.

```suggestion
  if (ruleResult && ruleResult.pass === false) {
    const message = (typeof ruleResult.message === 'function'
    	? ruleResult.message() : ruleResult.message) || `invalid ${typeof value} value`;
  	throw new Error(`[Enforce]/${rule.name}: ${message}`);
  }
}
```